File: src/main/java/wraith/fabricaeexnihilo/registry/barrel/LeakingRegistry.java
package wraith.fabricaeexnihilo.registry.barrel;

import alexiil.mc.lib.attributes.fluid.amount.FluidAmount;
import alexiil.mc.lib.attributes.fluid.volume.FluidVolume;
import com.google.common.reflect.TypeToken;
import net.minecraft.block.Block;
import net.minecraft.util.Pair;
import wraith.fabricaeexnihilo.FabricaeExNihilo;
import wraith.fabricaeexnihilo.api.recipes.barrel.LeakingRecipe;
import wraith.fabricaeexnihilo.api.registry.ILeakingRegistry;
import wraith.fabricaeexnihilo.compatibility.modules.MetaModule;
import wraith.fabricaeexnihilo.registry.AbstractRegistry;

import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

public class LeakingRegistry extends AbstractRegistry<List<LeakingRecipe>> implements ILeakingRegistry {

    private final List<LeakingRecipe> registry = new ArrayList<>();

    @Override
    public void clear() {
        registry.clear();
    }

    @Override
    protected List<LeakingRecipe> serializable() {
        return registry;
    }

    @Override
    public boolean register(LeakingRecipe recipe) {
        if(registry.stream().anyMatch(leakingRecipe -> leakingRecipe.fluid() == recipe.fluid() && leakingRecipe.target() == recipe.target())) {
            FabricaeExNihilo.LOGGER.warn("Conflicting Leaking Recipe not registered: " + recipe);
            return false;
        }
        return registry.add(recipe);
    }

    @Override
    public Pair<Block, FluidAmount> getResult(Block block, FluidVolume fluid) {
        var match = registry.stream().filter(recipe ->
                recipe.target().test(block) &&
                recipe.fluid().test(fluid) &&
                recipe.loss().isLessThanOrEqual(fluid.amount())).findFirst();
        if (match.isEmpty()) {
            return null;
        }
        var recipe = match.get();
        return new Pair<>(recipe.result(), recipe.loss());
    }

    @Override
    public List<LeakingRecipe> getREIRecipes() {
        return registry;
    }

    @Override
    public void registerJson(File file) {
        if (file.exists()){
            try (var reader = new FileReader(file)) {
                List<LeakingRecipe> json = gson.fromJson(
                        reader,
                        SERIALIZATION_TYPE
                );
                json.forEach(this::register);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    private static final Type SERIALIZATION_TYPE = new TypeToken<List<LeakingRecipe>>(){}.getType();

    public static ILeakingRegistry fromJson(File file) {
        return fromJson(file, LeakingRegistry::new, MetaModule.INSTANCE::registerLeaking);
    }

}

File: src/main/kotlin/exnihilofabrico/registry/barrel/LeakingRegistry.kt
package exnihilofabrico.registry.barrel

import alexiil.mc.lib.attributes.fluid.volume.FluidVolume
import com.google.gson.reflect.TypeToken
import exnihilofabrico.ExNihiloFabrico
import exnihilofabrico.api.recipes.barrel.LeakingRecipe
import exnihilofabrico.api.registry.ILeakingRegistry
import exnihilofabrico.compatibility.modules.MetaModule
import exnihilofabrico.registry.AbstractRegistry
import net.minecraft.block.Block
import java.io.File
import java.io.FileReader
import java.lang.reflect.Type

data class LeakingRegistry(val registry: MutableList<LeakingRecipe> = mutableListOf()):
    AbstractRegistry<MutableList<LeakingRecipe>>(), ILeakingRegistry {

    override fun serializable() = registry
    override fun clear() = registry.clear()

    override fun register(recipe: LeakingRecipe): Boolean {
        if(registry.any { it.fluid == recipe.fluid && it.target == recipe.target}) {
            ExNihiloFabrico.LOGGER.warn("Conflicting Leaking Recipe not registered: $recipe")
            return false
        }
        return registry.add(recipe)
    }

    override fun getResult(block: Block, fluid: FluidVolume): Pair<Block, Int>? {
        val match =
            registry.firstOrNull { it.target.test(block) && it.fluid.test(fluid) && it.loss <= fluid.amount } ?: return null
        return match.result to match.loss
    }

    override fun registerJson(file: File) {
        if(file.exists()){
            val json: MutableList<LeakingRecipe> = gson.fromJson(
                FileReader(file),
                SERIALIZATION_TYPE
            )
            json.forEach { register(it) }
        }
    }

    companion object {
        val SERIALIZATION_TYPE: Type? = object : TypeToken<MutableList<LeakingRecipe>>() {}.type
        fun fromJson(file: File) = fromJson(
            file,
            { LeakingRegistry() },
            MetaModule::registerLeaking
        )
    }

}